### ISSUE TITLE
auto-spy jest code issue - cannot find name jest

### ISSUE THREAD

**Issue by marimendez88** [gparlakov/scuri#107]
Following the steps you provide, when generating the auto-spy file

autoSpy<T> Function: 
` // turns out that in target:es2015 the methods attached to 

> the prototype are not enumerable so Object.keys returns []. So to workaround that and keep some backwards compatibility - merge with ownPropertyNames - that disregards the enumerable property.

  const keys = [...Object.keys(obj.prototype), ...Object.getOwnPropertyNames(obj.prototype)];
  keys.forEach(key => {
    res[key] = jest.fn();
  });`

VS code thowr an error saying: Cannot find name 'jest'
  res[key] = jest.fn();

same it `export type SpyOf<T> = T & {
  [k in keyof T]: T[k] extends (...args: any[]) => infer R ? T[k] & jest.Mock<R> : T[k];
};
`

Is there any fix, is something change? 

<img width="770" alt="Screen Shot 2020-12-08 at 15 13 16" src="https://user-images.githubusercontent.com/25093323/101542095-ea490200-3967-11eb-9136-8a761698b64f.png">

**Comment by gparlakov** [comment]
Are you using `jest` for testing? 

If not see the documentation for [autospy generation](https://github.com/gparlakov/scuri#autospy)

Specifically the commands for different setups and chose what you need.
angular | jest | jasmine | command
-- | -- | -- | --
5 |   | V | schematics scuri:autospy --legacy
5 | V |   | schematics scuri:autospy --for jest --legacy
6 |   | V | ng g scuri:autospy
6 | V |   | ng g scuri:autospy --for jest
7 |   | V | ng g scuri:autospy
7 | V |   | ng g scuri:autospy --for jest
8 |   | V | ng g scuri:autospy
8 | V |   | ng g scuri:autospy --for jest

Basically for angular 5 and below use `--legacy` and for `jasmine` do not add `--for jest`

**Comment by marimendez88** [comment]
yes @gparlakov , I am using jest for testing and the file is generated correctly, but inside of it, seems like the type SpyOf is not working and using a private name: 

![Uploading Screen Shot 2020-12-09 at 10.38.53.png…]()

**Comment by gparlakov** [comment]
It looks like the typing for jest is missing. 
How do you include the `@types/jest`?

**Comment by gparlakov** [comment]
@marimendez88 hi, did you manage to work out your issue? If so let's close this issue.

**Comment by venkatp57** [comment]
Hi, i am trying to use scuri in my project and i am getting some compilation issue on auto-spy.ts file

![image](https://user-images.githubusercontent.com/47427540/183692674-f7e2a1ea-af2a-464f-b6de-2bf4d3de7fd7.png)



can you please help on this.

**Comment by gparlakov** [comment]
Hi,

Sorry to hear that you're having trouble with Scuri.


From reading the TS error I can't get what it's complaining about. I'll
have to look deeper into it.

In the meantime: what does your environment look like? TS version angular
version node version and anything else that might help reproduce this?

**Comment by venkatp57** [comment]
Here are the details

"typescript": "~4.1.5",

![image](https://user-images.githubusercontent.com/47427540/183838185-b93ca130-842d-4213-bf22-27a523c64398.png)

**Comment by gparlakov** [comment]
Hey @venkatp57 I was able to reproduce the issue here -> [stackblitz link](https://stackblitz.com/edit/node-1sprrf?file=test-project%2Fauto-spy.ts,test-project%2Fsrc%2Fapp%2Fapp.component.ts,test-project%2Fsrc%2Fapp%2Ffoo.service.ts,test-project%2Ftsconfig.json,test-project%2Fsrc%2Fapp%2Fapp.component.spec.ts,package.json)

It boils down to telling typescript that we know what we are doing :)

Here's a version of autopsy that works as expected on the environment you specified 

```ts
/** Create an object with methods that are autoSpy-ed to use as mock dependency */
export function autoSpy<T>(obj: new (...args: any[]) => T): SpyOf<T> {
  const res: SpyOf<T> = {} as any;

  // turns out that in target:es2015 the methods attached to the prototype are not enumerable so Object.keys returns []. So to workaround that and keep some backwards compatibility - merge with ownPropertyNames - that disregards the enumerable property.
  // the Set remove duplicate entries
  const keys = new Set([
    ...(Object.keys(obj.prototype) as Array<keyof T>),
    ...(Object.getOwnPropertyNames(obj.prototype) as Array<keyof T>),
  ]);

  keys.forEach((key) => {
    if (typeof key === 'string') {
      (res[key] as any) = jasmine.createSpy(key);
    }
  });

  return res;
}

/** Keeps the types of properties of a type but assigns type of Spy to the methods */
export type SpyOf<T> = T &
  Partial<{
    [k in keyof T]: T[k] extends (...args: any[]) => any ? jasmine.Spy : T[k];
  }>;
```
